docs/provisioning.rst

```diff
@@ -0,0 +1,355 @@
+equivalent to the following stanza existing in the ``kernel.json`` file:
+
+.. code:: JSON
+
+      "metadata": {
+        "kernel_provisioner": {
+          "provisioner_name": "local-provisioner",
+          "config": {
+          }
+        }
+      },
+
+Should a given installation wish to use a *different* provisioner as
+their "default provisioner" (including subclasses of
+``LocalProvisioner``), they can do so by specifying a value for
+``KernelProvisionerFactory.default_provisioner_name``.
+
+Implementing a custom provisioner
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The impact of Kernel Provisioning is that it enables the ability to
+implement custom kernel provisioners to manage a kernel's lifecycle
+within any runtime environment. There are currently two approaches by
+which that can be accomplished, extending the ``KernelProvisionerBase``
+class or extending the built-in class - ``LocalProvisioner``. As more
+provisioners are introduced, some may be implemented in an abstract
+sense, from which specific implementations can be authored.
+
+Extending ``LocalProvisioner``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you're interested in running kernels locally and yet adjust their
+behavior, there's a good chance you can simply extend
+``LocalProvisioner`` via subclassing. This amounts to deriving from
+``LocalProvisioner`` and overriding appropriate methods to provide your
+custom functionality.
+
+In this example, RBACProvisioner will verify whether the current user is
+in the role meant for this kernel. If not, an exception will be thrown.
+
+.. code:: python
+
+    class RBACProvisioner(LocalProvisioner):
+
+        role: str = Unicode(config=True)
+
+        async def pre_launch(self, **kwargs: Any) -> Dict[str, Any]:
+
+            if not self.user_in_role(self.role):
+                raise PermissionError(f"User is not in role {self.role} and "
+                                      f"cannot launch this kernel.")
+
+            return super().pre_launch(**kwargs)
+
+It is important to note *when* it's necessary to call the superclass in
+a given method - since the operations it performs may be critical to the
+kernel's management. As a result, you'll likely need to become familiar
+with how ``LocalProvisioner`` operates.
+
+Extending ``KernelProvisionerBase``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If you'd like to launch your kernel in an environment other than the
+local server, then you will need to consider subclassing :class:`KernelProvisionerBase`
+directly.  This will allow you to implement the various kernel process
+controls relative to your target environment.  For instance, if you
+wanted to have your kernel hosted in a Hadoop YARN cluster, you will
+need to implement process-control methods like :meth:`poll` and :meth:`wait`
+to use the YARN REST API.  Or, similarly, a Kubernetes-based provisioner
+would need to implement the process-control methods using the Kubernetes client
+API, etc.
+
```

Comment:
Are thos `poll`, `wait`... lifecycle methods enforced by the API? Here also, would be good to have a list (or link) to the lifecycle methods (if the number is not too high and does not break the reading flow).

Reply:
Again, it's not clear to me what you mean - an example will help me.  I also don't understand what you mean by _enforced by the API_.  Sorry.

Reply:
I should look at the code but did not. My question is "Do all remote provisioners need to implement poll, wait...?". Another way to ask is "Are pool, wait... abstract methods that need to be implemented by all remote provisioners?

Reply:
Ah - yes, abstract methods are used where "enforcement" is required.  These methods are documented on [`KernelProvisionerBase`](https://github.com/kevin-bates/jupyter_client/blob/fc3227ee71ea7db8f54614d5eb58cec99bce98b5/jupyter_client/provisioning/provisioner_base.py#L24) and in the API portion of the provisioner docs.